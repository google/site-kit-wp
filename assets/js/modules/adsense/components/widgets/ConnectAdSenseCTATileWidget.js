/**
 * ConnectAdSenseCTATileWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import ConnectModuleCTATile from '../../../../components/KeyMetrics/ConnectModuleCTATile';
import useWidgetStateEffect from '../../../../googlesitekit/widgets/hooks/useWidgetStateEffect';

// Ensure the metadata instance is static when rerendering.
const metadata = { moduleSlug: 'adsense' };

export default function ConnectAdSenseCTATileWidget( { Widget, widgetSlug } ) {
	useWidgetStateEffect( widgetSlug, ConnectModuleCTATile, metadata );

	// Note that we need to render the `Widget` component as a wrapper here so this component will display
	// correctly when used as a `FallbackComponent` for the `whenActive` HOC. Conversely, when `ConnectModuleCTATile`
	// is rendered as an `OverrideComponent` in `WidgetRenderer` (as a result of multiple adjacent widgets rendering
	// this component and thus sharing the same state), it is wrapped with a `Widget` component - the net
	// result being the ConnectModuleCTATile is correctly wrapped in a `Widget` in both cases.
	return (
		<Widget>
			<ConnectModuleCTATile { ...metadata } />
		</Widget>
	);
}

ConnectAdSenseCTATileWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	widgetSlug: PropTypes.string.isRequired,
};
