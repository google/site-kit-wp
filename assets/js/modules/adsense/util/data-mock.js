/**
 * AdSense test data mock.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import faker from 'faker';
import invariant from 'invariant';
import md5 from 'md5';
import { castArray, isPlainObject } from 'lodash';
import { range } from 'rxjs';
import { map, reduce } from 'rxjs/operators';

/**
 * Internal dependencies
 */
import { MODULES_ADSENSE } from '../datastore/constants';
import { getDateString, isValidDateString, stringToDate } from '../../../util';
import { validateMetrics } from './report-validation';
import { dateInstanceToAdSenseDate } from './date';

const METRIC_RATIO = 'METRIC_RATIO';
const METRIC_TALLY = 'METRIC_TALLY';
const METRIC_CURRENCY = 'METRIC_CURRENCY';

const ADSENSE_METRIC_TYPES = {
	ESTIMATED_EARNINGS: METRIC_CURRENCY,
	PAGE_VIEWS_RPM: METRIC_CURRENCY,
	IMPRESSIONS: METRIC_TALLY,
	PAGE_VIEWS_CTR: METRIC_RATIO,
};

class DataFactory {
	/**
	 * Constructor.
	 *
	 * @since 1.36.0
	 */
	constructor() {
		this.metricMemory = {};
	}

	/**
	 * Generates and returns report headers.
	 *
	 * @since 1.36.0
	 *
	 * @param {Array.<string>} metrics    Metrics list.
	 * @param {Array.<string>} dimensions Dimensions list.
	 * @return {Array.<Object>} Headers list.
	 */
	createHeaders( metrics, dimensions ) {
		const headers = [];

		dimensions.forEach( ( dimension ) => {
			headers.push( {
				currencyCode: null,
				name: dimension.toUpperCase(),
				type: 'DIMENSION',
			} );
		} );

		metrics.forEach( ( metric ) => {
			const ucMetric = metric.toUpperCase();
			headers.push( {
				currencyCode:
					ADSENSE_METRIC_TYPES[ ucMetric ] === METRIC_CURRENCY
						? 'USD'
						: null,
				name: ucMetric,
				type: ADSENSE_METRIC_TYPES[ ucMetric ],
			} );
		} );

		return headers;
	}

	/**
	 * Generates and returns metric values.
	 *
	 * @since 1.36.0
	 *
	 * @param {string}         date    The current date.
	 * @param {Array.<string>} metrics Metrics list.
	 * @return {Array.<string>} Array of metric values.
	 */
	createMetricValues( date, metrics ) {
		const values = [];
		const delta = 0.15;

		for ( const metric of metrics ) {
			const lastValue = this.metricMemory[ metric ];
			const min = lastValue * ( 1 - delta );
			const max = lastValue * ( 1 + delta );

			switch ( ADSENSE_METRIC_TYPES[ metric.toUpperCase() ] ) {
				case METRIC_TALLY: {
					const options =
						lastValue === undefined
							? { min: 500, max: 700 }
							: {
									min: Math.floor( min ),
									max: Math.ceil( max ),
							  };

					const newValue = faker.datatype.number( options );
					values.push( newValue.toString() );
					this.metricMemory[ metric ] = newValue;
					break;
				}
				case METRIC_CURRENCY: {
					const options =
						lastValue === undefined
							? { min: 500, max: 700 }
							: { min, max };

					const newValue = faker.datatype.float( options );
					values.push( newValue.toFixed( 2 ) );
					this.metricMemory[ metric ] = newValue;
					break;
				}
				case METRIC_RATIO: {
					const options =
						lastValue === undefined
							? { min: 0.4, max: 0.6 }
							: {
									min,
									max: Math.min( max, 1 ),
							  };

					const newValue = faker.datatype.float( options );
					values.push( newValue.toFixed( 2 ) );
					this.metricMemory[ metric ] = newValue;
					break;
				}
				default:
					values.push( '' );
					break;
			}
		}

		return values;
	}

	/**
	 * Generates and returns dimension values.
	 *
	 * @since 1.36.0
	 *
	 * @param {string}         date       The current date.
	 * @param {Array.<string>} dimensions Dimensions list.
	 * @return {Array.<string>} Array of dimension values.
	 */
	createDimensionValues( date, dimensions ) {
		const values = [];

		for ( const dimension of dimensions ) {
			switch ( dimension.toUpperCase() ) {
				case 'DATE':
					values.push( date );
					break;
			}
		}

		return values;
	}
}

/**
 * Creates a row object from an array of values.
 *
 * @since 1.36.0
 *
 * @param {Array} array Array.
 * @return {Object} Row.
 */
function rowFromArray( array ) {
	return {
		cells: array.map( ( value ) => ( { value } ) ),
	};
}

/**
 * Generates mock data for AdSense reports.
 *
 * @since 1.36.0
 *
 * @param {Object} args Report options.
 * @return {Array.<Object>} An array with generated report.
 */
export function getAdSenseMockResponse( args ) {
	invariant(
		isPlainObject( args ),
		'report options are required to generate a mock response.'
	);
	invariant(
		isValidDateString( args.startDate ),
		'a valid startDate is required.'
	);
	invariant(
		isValidDateString( args.endDate ),
		'a valid endDate is required.'
	);
	validateMetrics( args.metrics );

	const originalSeedValue = faker.seedValue;
	const argsHash = parseInt(
		md5( JSON.stringify( args ) ).substring( 0, 10 ),
		16
	);

	// We set seed for every data mock to make sure that the same arguments get the same report data.
	// It means that everyone will have the same report data and will see the same widgets in the storybook.
	// This approach gives us additional flexibility to control randomness on a per widget basis.
	if ( ! Number.isNaN( argsHash ) ) {
		faker.seed( argsHash );
	}

	const factory = new DataFactory();
	const metrics = castArray( args.metrics ).filter( ( metric ) => !! metric );
	const dimensions = castArray( args.dimensions ).filter(
		( dimension ) => !! dimension
	);

	const data = {
		warnings: [],
		startDate: args.startDate,
		endDate: args.endDate,
		totalMatchedRows: '0',
		headers: factory.createHeaders( metrics, dimensions ),
		totals: [],
		averages: [],
		rows: [],
	};

	const startDate = stringToDate( args.startDate );
	const endDate = stringToDate( args.endDate );
	const dayInMilliseconds = 24 * 60 * 60 * 1000;
	const totalDays = 1 + ( endDate - startDate ) / dayInMilliseconds; // +1 to include the endDate into the dates range.

	// This is the list of operations that we will apply to the range (array) of numbers.
	const ops = [
		// Converts range number to a date string.
		map( ( item ) => {
			// Valid use of `new Date()` with an argument.
			// eslint-disable-next-line sitekit/no-direct-date
			const updatedMilliseconds = new Date( startDate ).setDate(
				startDate.getDate() + item
			);

			// Valid use of `new Date()` with an argument.
			// eslint-disable-next-line sitekit/no-direct-date
			return getDateString( new Date( updatedMilliseconds ) );
		} ),
		// Add dimension and metric values.
		map( ( date ) => [
			...factory.createDimensionValues( date, dimensions ),
			...factory.createMetricValues( date, metrics ),
		] ),
		// Accumulate all rows into a single array.
		reduce( ( rows, row ) => [ ...rows, row ], [] ),
	];

	// Process the stream of dates and add generated rows to the report data object.
	range( 0, totalDays )
		.pipe( ...ops )
		.subscribe( ( rows ) => {
			data.rows = rows;
			data.totalMatchedRows = rows.length.toString();

			// We pretend that the first row contains averages and the last one totals because we don't
			// really need mathematically correct values and can simplify the process of finding this information.
			data.totals = [ ...( rows[ rows.length - 1 ] || [] ) ];
			data.averages = [ ...( rows[ 0 ] || [] ) ];
		} );

	// Set the original seed value for the faker.
	faker.seed( originalSeedValue );

	return {
		...data,
		// v2 transforms.
		rows: data.rows.map( ( rowArray ) => rowFromArray( rowArray ) ),
		totals: rowFromArray( data.totals ),
		averages: rowFromArray( data.averages ),
		startDate: dateInstanceToAdSenseDate( startDate ),
		endDate: dateInstanceToAdSenseDate( endDate ),
	};
}

/**
 * Generates mock response for AdSense reports.
 *
 * @since 1.36.0
 *
 * @param {wp.data.registry} registry Registry with all available stores registered.
 * @param {Object}           options  Report options.
 */
export function provideAdSenseMockReport( registry, options ) {
	registry
		.dispatch( MODULES_ADSENSE )
		.receiveGetReport( getAdSenseMockResponse( options ), {
			options,
		} );
}
