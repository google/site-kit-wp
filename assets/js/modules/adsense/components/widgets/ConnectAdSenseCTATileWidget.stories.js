/**
 * ConnectAdSenseCTATileWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import {
	provideModuleRegistrations,
	provideModules,
} from '../../../../../../tests/js/utils';
import { provideKeyMetricsWidgetRegistrations } from '../../../../components/KeyMetrics/test-utils';
import { KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT } from '../../../../googlesitekit/datastore/user/constants';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import ConnectAdSenseCTATileWidget from './ConnectAdSenseCTATileWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectAdSenseCTATile'
)( ConnectAdSenseCTATileWidget );

function Template() {
	return <WidgetWithComponentProps />;
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectAdSenseCTATileWidget';
Default.args = {
	keyMetricsWidgets: {
		[ KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT ]: {
			modules: [ 'adsense' ],
		},
		secondAdSenseWidget: {
			modules: [ 'adsense' ],
		},
	},
};
Default.scenario = {
	// eslint-disable-next-line sitekit/no-storybook-scenario-label
	delay: 250,
};

export const WithSingleWidget = Template.bind( {} );
WithSingleWidget.storyName = 'ConnectAdSenseCTATileWidget (for single widget)';
WithSingleWidget.args = {
	keyMetricsWidgets: {
		[ KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT ]: {
			modules: [ 'adsense' ],
		},
	},
};
WithSingleWidget.scenario = {};

export default {
	title: 'Key Metrics/ConnectAdSenseCTATileWidget',
	decorators: [
		( Story, { args } ) => {
			const setupRegistry = ( registry ) => {
				provideModules( registry, [
					{
						slug: 'adsense',
						active: false,
						connected: false,
					},
				] );
				provideModuleRegistrations( registry );
				provideKeyMetricsWidgetRegistrations(
					registry,
					args?.keyMetricsWidgets
				);
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
