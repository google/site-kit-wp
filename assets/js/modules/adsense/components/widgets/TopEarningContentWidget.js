/**
 * TopEarningContentWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { compose } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { useSelect, useInViewSelect } from 'googlesitekit-data';
import {
	CORE_USER,
	KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT,
} from '../../../../googlesitekit/datastore/user/constants';
import { DATE_RANGE_OFFSET, MODULES_ADSENSE } from '../../datastore/constants';
import {
	MetricTileTable,
	MetricTileTablePlainText,
} from '../../../../components/KeyMetrics';
import Link from '../../../../components/Link';
import { ZeroDataMessage } from '../../../analytics-4/components/common';
import { numFmt } from '../../../../util';
import whenActive from '../../../../util/when-active';
import ConnectGA4CTATileWidget from '../../../analytics-4/components/widgets/ConnectGA4CTATileWidget';
import useViewOnly from '../../../../hooks/useViewOnly';
import { AdSenseLinkCTA } from '../common';
import { MODULES_ANALYTICS_4 } from '../../../analytics-4/datastore/constants';
import ConnectAdSenseCTATileWidget from './ConnectAdSenseCTATileWidget';

function TopEarningContentWidget( { Widget } ) {
	const viewOnlyDashboard = useViewOnly();

	const dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
		} )
	);

	const adSenseAccountID = useSelect( ( select ) =>
		select( MODULES_ADSENSE ).getAccountID()
	);

	const reportOptions = {
		...dates,
		dimensions: [ 'pagePath', 'adSourceName' ],
		metrics: [ { name: 'totalAdRevenue' } ],
		dimensionFilters: {
			adSourceName: `Google AdSense account (${ adSenseAccountID })`,
		},
		orderby: [
			{
				metric: { metricName: 'totalAdRevenue' },
				desc: true,
			},
		],
		limit: 3,
	};

	const report = useInViewSelect(
		( select ) => {
			return select( MODULES_ANALYTICS_4 ).getReport( reportOptions );
		},
		[ reportOptions ]
	);

	const error = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getErrorForSelector( 'getReport', [
			reportOptions,
		] )
	);

	const titles = useInViewSelect(
		( select ) =>
			! error
				? select( MODULES_ANALYTICS_4 ).getPageTitles(
						report,
						reportOptions
				  )
				: undefined,
		[ report, reportOptions ]
	);

	const loading = useSelect(
		( select ) =>
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ reportOptions ]
			) || titles === undefined
	);

	const isAdSenseLinked = useSelect( ( select ) => {
		if ( viewOnlyDashboard && loading ) {
			return undefined;
		}

		return select( MODULES_ANALYTICS_4 ).getAdSenseLinked();
	} );

	if ( ! isAdSenseLinked && ! viewOnlyDashboard ) {
		return (
			<Widget>
				<AdSenseLinkCTA />
			</Widget>
		);
	}

	const { rows = [] } = report || {};

	const columns = [
		{
			field: 'dimensionValues.0.value',
			Component( { fieldValue } ) {
				const url = fieldValue;
				const title = titles[ url ];
				// Utilizing `useSelect` inside the component rather than
				// returning its direct value to the `columns` array.
				// This pattern ensures that the component re-renders correctly based on changes in state,
				// preventing potential issues with stale or out-of-sync data.
				// Note: This pattern is replicated in a few other spots within our codebase.
				const serviceURL = useSelect( ( select ) => {
					return ! viewOnlyDashboard
						? select( MODULES_ANALYTICS_4 ).getServiceReportURL(
								'all-pages-and-screens',
								{
									filters: {
										unifiedPagePathScreen: url,
									},
									dates,
								}
						  )
						: null;
				} );

				if ( viewOnlyDashboard ) {
					return <MetricTileTablePlainText content={ title } />;
				}

				return (
					<Link
						href={ serviceURL }
						title={ title }
						external
						hideExternalIndicator
					>
						{ title }
					</Link>
				);
			},
		},
		{
			field: 'metricValues.0.value',
			Component( { fieldValue } ) {
				return (
					<strong>
						{ numFmt( fieldValue, {
							style: 'currency',
							currency: report?.metadata?.currencyCode,
						} ) }
					</strong>
				);
			},
		},
	];

	return (
		<MetricTileTable
			Widget={ Widget }
			widgetSlug={ KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT }
			loading={ loading }
			rows={ rows }
			columns={ columns }
			ZeroState={ ZeroDataMessage }
			error={ error }
			moduleSlug="analytics-4"
		/>
	);
}

TopEarningContentWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};

export default compose(
	whenActive( {
		moduleName: 'analytics-4',
		FallbackComponent: ConnectGA4CTATileWidget,
	} ),
	whenActive( {
		moduleName: 'adsense',
		FallbackComponent: ConnectAdSenseCTATileWidget,
	} )
)( TopEarningContentWidget );
