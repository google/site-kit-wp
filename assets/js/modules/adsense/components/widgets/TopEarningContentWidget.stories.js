/**
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import {
	provideKeyMetrics,
	provideModules,
} from '../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import { replaceValuesInAnalytics4ReportWithZeroData } from '../../../../../../.storybook/utils/zeroReports';
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import { Provider as ViewContextProvider } from '../../../../components/Root/ViewContextContext';
import {
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
} from '../../../../googlesitekit/constants';
import TopEarningContentWidget from './TopEarningContentWidget';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../../../util/errors';
import {
	STRATEGY_ZIP,
	getAnalytics4MockResponse,
	provideAnalytics4MockReport,
} from '../../../analytics-4/utils/data-mock';
import { MODULES_ANALYTICS_4 } from '../../../analytics-4/datastore/constants';
import { MODULES_ADSENSE } from '../../datastore/constants';

const adSenseAccountID = 'pub-1234567890';

const reportOptions = {
	startDate: '2020-08-11',
	endDate: '2020-09-07',
	dimensions: [ 'pagePath', 'adSourceName' ],
	metrics: [ { name: 'totalAdRevenue' } ],
	dimensionFilters: {
		adSourceName: `Google AdSense account (${ adSenseAccountID })`,
	},
	orderby: [
		{
			metric: { metricName: 'totalAdRevenue' },
			desc: true,
		},
	],
	limit: 3,
};

const pageTitlesReportOptions = {
	startDate: '2020-08-11',
	endDate: '2020-09-07',
	dimensionFilters: {
		pagePath: new Array( 3 )
			.fill( '' )
			.map( ( _, i ) => `/test-post-${ i + 1 }/` )
			.sort(),
	},
	dimensions: [ 'pagePath', 'pageTitle' ],
	metrics: [ { name: 'screenPageViews' } ],
	orderby: [ { metric: { metricName: 'screenPageViews' }, desc: true } ],
	limit: 15,
};

const WidgetWithComponentProps = withWidgetComponentProps( 'test' )(
	TopEarningContentWidget
);

function Template( { setupRegistry, viewContext, ...args } ) {
	return (
		<WithRegistrySetup func={ setupRegistry }>
			<ViewContextProvider
				value={ viewContext || VIEW_CONTEXT_MAIN_DASHBOARD }
			>
				<WidgetWithComponentProps { ...args } />
			</ViewContextProvider>
		</WithRegistrySetup>
	);
}

export const Ready = Template.bind( {} );
Ready.storyName = 'Ready';
Ready.args = {
	setupRegistry: ( registry ) => {
		const pageTitlesReport = getAnalytics4MockResponse(
			pageTitlesReportOptions,
			// Use the zip combination strategy to ensure a one-to-one mapping of page paths to page titles.
			// Otherwise, by using the default cartesian product of dimension values, the resulting output will have non-matching
			// page paths to page titles.
			{ dimensionCombinationStrategy: STRATEGY_ZIP }
		);

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetReport( pageTitlesReport, {
				options: pageTitlesReportOptions,
			} );

		provideAnalytics4MockReport( registry, reportOptions );
	},
};
Ready.scenario = {
	label: 'KeyMetrics/TopEarningContentWidget/Ready',
	delay: 250,
};

export const ReadyViewOnly = Template.bind( {} );
ReadyViewOnly.storyName = 'Ready View Only';
ReadyViewOnly.args = {
	setupRegistry: ( registry ) => {
		const pageTitlesReport = getAnalytics4MockResponse(
			pageTitlesReportOptions,
			{ dimensionCombinationStrategy: STRATEGY_ZIP }
		);

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetReport( pageTitlesReport, {
				options: pageTitlesReportOptions,
			} );

		provideAnalytics4MockReport( registry, reportOptions );
	},
	viewContext: VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
};
ReadyViewOnly.scenario = {
	label: 'KeyMetrics/TopEarningContentWidget/ReadyViewOnly',
	delay: 250,
};

export const Loading = Template.bind( {} );
Loading.storyName = 'Loading';
Loading.args = {
	setupRegistry: ( { dispatch } ) => {
		dispatch( MODULES_ANALYTICS_4 ).startResolution( 'getReport', [
			reportOptions,
		] );
	},
};

export const ZeroData = Template.bind( {} );
ZeroData.storyName = 'Zero Data';
ZeroData.args = {
	setupRegistry: ( { dispatch } ) => {
		const report = getAnalytics4MockResponse( reportOptions );
		const zeroReport =
			replaceValuesInAnalytics4ReportWithZeroData( report );

		dispatch( MODULES_ANALYTICS_4 ).receiveGetReport( zeroReport, {
			options: reportOptions,
		} );
	},
};

export const Error = Template.bind( {} );
Error.storyName = 'Error';
Error.args = {
	setupRegistry: ( { dispatch } ) => {
		const errorObject = {
			code: 400,
			message: 'Test error message. ',
			data: {
				status: 400,
				reason: 'badRequest',
			},
		};

		dispatch( MODULES_ANALYTICS_4 ).receiveError(
			errorObject,
			'getReport',
			[ reportOptions ]
		);

		dispatch( MODULES_ANALYTICS_4 ).finishResolution( 'getReport', [
			reportOptions,
		] );
	},
};

export const InsufficientPermissions = Template.bind( {} );
InsufficientPermissions.storyName = 'Insufficient Permissions';
InsufficientPermissions.args = {
	setupRegistry: ( { dispatch } ) => {
		const errorObject = {
			code: 403,
			message: 'Test error message. ',
			data: {
				status: 403,
				reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS,
			},
		};

		dispatch( MODULES_ANALYTICS_4 ).receiveError(
			errorObject,
			'getReport',
			[ reportOptions ]
		);

		dispatch( MODULES_ANALYTICS_4 ).finishResolution( 'getReport', [
			reportOptions,
		] );
	},
};

export const AdSenseNotLinked = Template.bind( {} );
AdSenseNotLinked.storyName = 'AdSense Not Linked';
AdSenseNotLinked.args = {
	setupRegistry: ( registry ) => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setAdSenseLinked( false );
	},
};
AdSenseNotLinked.scenario = {
	label: 'KeyMetrics/TopEarningContentWidget/AdSenseNotLinked',
	delay: 250,
};

export default {
	title: 'Key Metrics/TopEarningContentWidget',
	decorators: [
		( Story, { args } ) => {
			const setupRegistry = ( registry ) => {
				provideModules( registry, [
					{
						slug: 'adsense',
						active: true,
						connected: true,
					},
					{
						slug: 'analytics-4',
						active: true,
						connected: true,
					},
				] );

				registry.dispatch( CORE_USER ).setReferenceDate( '2020-09-08' );

				provideKeyMetrics( registry );

				registry
					.dispatch( MODULES_ANALYTICS_4 )
					.setAdSenseLinked( true );

				registry
					.dispatch( MODULES_ADSENSE )
					.setAccountID( adSenseAccountID );

				// Call story-specific setup.
				args.setupRegistry( registry );
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
