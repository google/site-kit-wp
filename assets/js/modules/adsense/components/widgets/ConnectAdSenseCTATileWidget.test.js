/**
 * ConnectAdSenseCTATileWidget component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	provideModules,
	provideUserCapabilities,
	render,
} from '../../../../../../tests/js/test-utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import ConnectAdSenseCTATileWidget from './ConnectAdSenseCTATileWidget';

describe( 'ConnectAdSenseCTATileWidget', () => {
	const WidgetWithComponentProps = withWidgetComponentProps(
		'keyMetricsConnectAdSenseCTATile'
	)( ConnectAdSenseCTATileWidget );

	it( 'should render the Connect AdSense CTA tile', () => {
		const registry = createTestRegistry();
		provideUserCapabilities( registry );
		provideModules( registry, [
			{
				slug: 'adsense',
				active: false,
				connected: false,
			},
		] );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		expect( container ).toMatchSnapshot();

		expect( container ).toHaveTextContent( 'Connect AdSense' );
	} );
} );
