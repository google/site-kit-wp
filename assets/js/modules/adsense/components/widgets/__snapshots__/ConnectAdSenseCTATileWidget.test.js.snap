// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectAdSenseCTATileWidget should render a title when only a single Connect AdSense CTA is present 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--keyMetricsConnectAdSenseCTATile"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-widget--connectModuleCTATile googlesitekit-km-widget-tile false"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-typography googlesitekit-km-widget-tile__title googlesitekit-typography--label googlesitekit-typography--small"
          >
            Top earning pages
          </h3>
          <span
            class="googlesitekit-info-tooltip"
            title="Pages that generated the most AdSense revenue"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-connect-module-cta-tile"
          >
            <div
              class="googlesitekit-km-connect-module-cta-tile__content"
            >
              <p
                class="googlesitekit-km-connect-module-cta-tile__text"
              >
                AdSense is disconnected, metric can’t be displayed
              </p>
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Connect AdSense
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConnectAdSenseCTATileWidget should render the Connect AdSense CTA tile 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--keyMetricsConnectAdSenseCTATile"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-widget--connectModuleCTATile googlesitekit-km-widget-tile  googlesitekit-km-widget-tile--combined"
      >
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-connect-module-cta-tile"
          >
            <div
              class="googlesitekit-km-connect-module-cta-tile__content"
            >
              <p
                class="googlesitekit-km-connect-module-cta-tile__text"
              >
                AdSense is disconnected, some of your metrics can’t be displayed
              </p>
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Connect AdSense
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          class="googlesitekit-km-connect-module-cta-tile__ghost-card"
        >
          <div
            class="googlesitekit-preview-block"
            style="width: 100%; height: 90px;"
          >
            <div
              class="googlesitekit-preview-block__wrapper"
            />
          </div>
        </div>
        <div
          class="googlesitekit-km-connect-module-cta-tile__ghost-card"
        >
          <div
            class="googlesitekit-preview-block"
            style="width: 100%; height: 90px;"
          >
            <div
              class="googlesitekit-preview-block__wrapper"
            />
          </div>
        </div>
        <div
          class="googlesitekit-km-connect-module-cta-tile__ghost-card"
        >
          <div
            class="googlesitekit-preview-block"
            style="width: 100%; height: 90px;"
          >
            <div
              class="googlesitekit-preview-block__wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
